orttraining/orttraining/test/graph/gradient_graph_builder_test.cc

```diff
@@ -1115,7 +1115,113 @@ void RetrieveSendRecvOperators(
   }
 }
 
-TEST(GradientGraphBuilderTest, PipelineOnlinePartition) {
+TEST(GradientGraphBuilderTest, PipelineOnlinePartition_bert_tiny) {
+  const auto model_path = ORT_TSTR("testdata/bert-tiny.onnx");
+
+  const size_t total_partition_count = 3;
+  TrainingSession::TrainingConfiguration::PipelineConfiguration pipe{};
+  pipe.do_partition = true;
+
+  // evenly model in 3 partitions
+  TrainingSession::TrainingConfiguration::CutInfo cut0 = {
+      onnxruntime::training::TrainingSession::TrainingConfiguration::CutEdge("186"),
+      onnxruntime::training::TrainingSession::TrainingConfiguration::CutEdge("71", {"273"})};
+
+  TrainingSession::TrainingConfiguration::CutInfo cut1 = {
+      onnxruntime::training::TrainingSession::TrainingConfiguration::CutEdge("308"),
+      onnxruntime::training::TrainingSession::TrainingConfiguration::CutEdge("71", {"395"})};
+
+  pipe.cut_list.emplace_back(cut0);
+  pipe.cut_list.emplace_back(cut1);
+
+  TrainingSession::TrainingConfiguration::MixedPrecisionConfiguration mixed_precision_config{};
+  mixed_precision_config.use_fp16_initializers = true;
+
+  // 2 test variations - full precision and mixed precision
+  const std::vector<bool> test_with_fp32{true, false};
+  for (auto is_fp32 : test_with_fp32) {
+    // graph is partitioned into 4 parts.
+    for (int i = 0; i < static_cast<int>(total_partition_count); ++i) {
+#ifdef _WIN32
+      auto surfix = std::to_wstring(i);
+#else
+      auto surfix = std::to_string(i);
+#endif
+      PathString output_file = ORT_TSTR("pipeline_partition_") + surfix + ORT_TSTR("_back.onnx");
+
+      auto config = MakeBasicTrainingConfig();
+
+      if (i == static_cast<int>(total_partition_count - 1)) {
+        config.loss_function_config = TrainingSession::TrainingConfiguration::LossFunctionConfiguration{};
+        config.loss_function_config.value().loss_function_info =
+            LossFunctionInfo(OpDef("BertLoss", kOnnxDomain),
+                             "total_loss",
+                             {/*prediction_masked_lm*/ "output1",
+                              /*prediction_next_sentence*/ "output2",
+                              /*masked_lm_positions*/ "masked_lm_positions",
+                              /*masked_lm_ids*/ "masked_lm_ids",
+                              /*masked_lm_weights*/ "masked_lm_weights",
+                              /*next_sentence_labels*/ "next_sentence_labels",
+                              /*mlm_loss*/ "mlm_loss",
+                              /*nsp_loss*/ "nsp_loss"});
+      }
+
+      config.immutable_weights = {
+          {"Div", {{1, 8.0f}, {1, 1.4142135381698608f}}},
+          {"Add", {{1, 1.0f}, {1, 9.999999960041972e-13f}}},
+          {"Mul", {{1, 0.5f}, {1, -10000.0f}}},
+          {"Sub", {{0, 1.0f}}}};
```

Comment:
What are these magic numbers? (I see similar code elsewhere in the codebase, but I'm not sure what they have to do with the pipelining code....)

Are these related to initializers, or to `weight_names_not_to_train` in the `TrainingConfiguration`? Do we/should we have a test for the case where two pipeline stages depend on the same immutable weight? (Presumably in that case we'd replicate the initializer rather than sending it down the pipeline on each step.)

Reply:
Those are configurations to train bert model. More specifically, some weights in bert model that we treat as constant and won't update during training.

It is needed for training bert (like in the test TrainingSession_BertToy). But I guess in this test, since we are just doing partition, we don't need to include it. I'll remove it to make the it cleaner.

---
In reply to: [449312077](https://github.com/microsoft/onnxruntime/pull/4321#discussion_r449312077) [](ancestors = 449312077)